### PR TITLE
Visualize training of 2d TBoxes

### PR DESCRIPTION
**Pull request by brozonoyer**
cc @zhangdongxu @dhruvdcoder

### CODE REVIEW

**Comment by brozonoyer** [comment]
@dhruvdcoder here is my implementation of exact negative sampling, using every negative candidate exactly once and masking the remaining negative edges in each batch. This hasn't improved the box visualization massively so I would appreciate a second pair of eyes to check if there is anything wrong with the implementation (esp. in the changes to the loss function with masked softmax).

cc @mboratko @zhangdongxu
